File: test/javascript/form_validation_test.js
// Test file for form validation functionality
// Note: This would typically be run with a JavaScript test runner like Jest or similar

// Mock DOM elements for testing (would be replaced with proper test setup)
function createMockForm() {
  const form = document.createElement('div');
  form.setAttribute('data-controller', 'form-validation');
  form.setAttribute('data-form-validation-mode-value', 'registration');
  
  const emailField = document.createElement('input');
  emailField.type = 'email';
  emailField.setAttribute('data-form-validation-target', 'emailField');
  
  const passwordField = document.createElement('input');
  passwordField.type = 'password';
  passwordField.setAttribute('data-form-validation-target', 'passwordField');
  
  const passwordConfirmationField = document.createElement('input');
  passwordConfirmationField.type = 'password';
  passwordConfirmationField.setAttribute('data-form-validation-target', 'passwordConfirmationField');
  
  const submitButton = document.createElement('button');
  submitButton.setAttribute('data-form-validation-target', 'submitButton');
  
  const emailError = document.createElement('div');
  emailError.className = 'hidden';
  emailError.setAttribute('data-form-validation-target', 'emailError');
  
  const passwordError = document.createElement('div');
  passwordError.className = 'hidden';
  passwordError.setAttribute('data-form-validation-target', 'passwordError');
  
  const passwordConfirmationError = document.createElement('div');
  passwordConfirmationError.className = 'hidden';
  passwordConfirmationError.setAttribute('data-form-validation-target', 'passwordConfirmationError');
  
  const passwordStrength = document.createElement('div');
  passwordStrength.className = 'hidden';
  passwordStrength.setAttribute('data-form-validation-target', 'passwordStrength');
  
  form.appendChild(emailField);
  form.appendChild(passwordField);
  form.appendChild(passwordConfirmationField);
  form.appendChild(submitButton);
  form.appendChild(emailError);
  form.appendChild(passwordError);
  form.appendChild(passwordConfirmationError);
  form.appendChild(passwordStrength);
  
  return {
    form,
    emailField,
    passwordField,
    passwordConfirmationField,
    submitButton,
    emailError,
    passwordError,
    passwordConfirmationError,
    passwordStrength
  };
}

// Test cases (would be wrapped in proper test framework)
console.log('Form validation tests - these verify the expected behavior:');

console.log('1. Email validation:');
console.log('   - Empty email should clear validation');
console.log('   - Invalid email should show error');
console.log('   - Valid email should show success');

console.log('2. Password validation:');
console.log('   - Empty password should clear validation');
console.log('   - Password < 6 chars should show error');
console.log('   - Password >= 6 chars should show success');

console.log('3. Password strength:');
console.log('   - < 8 chars = weak');
console.log('   - 8+ chars with letters and numbers = medium');
console.log('   - 8+ chars with letters, numbers, and special chars = strong');

console.log('4. Password confirmation:');
console.log('   - Empty confirmation should clear validation');
console.log('   - Mismatched passwords should show error');
console.log('   - Matched passwords should show success');

console.log('5. Form validation:');
console.log('   - Submit button disabled when form invalid');
console.log('   - Submit button enabled when form valid');

// Export for potential use in actual test runners
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    createMockForm
  };
}

File: test/integration/form_validation_test.rb
require "test_helper"

class FormValidationTest < ActionDispatch::IntegrationTest
  test "registration form has validation elements" do
    get sign_up_path
    assert_response :success
    
    # Check that the form has the validation controller attached
    assert_select "div[data-controller='form-validation']"
    assert_select "div[data-form-validation-mode-value='registration']"
    
    # Check that email field has validation targets and actions
    assert_select "input[data-form-validation-target='emailField']"
    assert_select "input[data-action*='form-validation#validateEmail']"
    
    # Check that password field has validation targets and actions
    assert_select "input[data-form-validation-target='passwordField']"
    assert_select "input[data-action*='form-validation#validatePassword']"
    
    # Check that password confirmation field has validation targets and actions
    assert_select "input[data-form-validation-target='passwordConfirmationField']"
    assert_select "input[data-action*='form-validation#validatePasswordConfirmation']"
    
    # Check that submit button has validation target
    assert_select "input[data-form-validation-target='submitButton']"
    
    # Check that error message containers exist
    assert_select "div[data-form-validation-target='emailError']"
    assert_select "div[data-form-validation-target='passwordError']"
    assert_select "div[data-form-validation-target='passwordConfirmationError']"
    assert_select "div[data-form-validation-target='passwordStrength']"
  end
  
  test "login form has validation elements" do
    get new_session_path
    assert_response :success
    
    # Check that the form has the validation controller attached
    assert_select "div[data-controller='form-validation']"
    assert_select "div[data-form-validation-mode-value='login']"
    
    # Check that email field has validation targets and actions
    assert_select "input[data-form-validation-target='emailField']"
    assert_select "input[data-action*='form-validation#validateEmail']"
    
    # Check that password field has validation targets and actions
    assert_select "input[data-form-validation-target='passwordField']"
    assert_select "input[data-action*='form-validation#validatePassword']"
    
    # Check that submit button has validation target
    assert_select "input[data-form-validation-target='submitButton']"
    
    # Check that error message containers exist
    assert_select "div[data-form-validation-target='emailError']"
    assert_select "div[data-form-validation-target='passwordError']"
    
    # Check that password strength is NOT present in login form
    assert_select "div[data-form-validation-target='passwordStrength']", false
    assert_select "div[data-form-validation-target='passwordConfirmationError']", false
  end
  
  test "form fields have proper CSS classes for validation states" do
    get sign_up_path
    assert_response :success
    
    # Check that fields have the border-gray-400 class initially
    assert_select "input.border-gray-400"
    
    # Check that fields have transition classes
    assert_select "input.transition-colors"
    
    # Check that submit button has disabled state classes
    assert_select "input.disabled\\:opacity-50"
    assert_select "input.disabled\\:cursor-not-allowed"
    
    # Check that error containers are initially hidden
    assert_select "div.hidden[data-form-validation-target='emailError']"
    assert_select "div.hidden[data-form-validation-target='passwordError']"
    assert_select "div.hidden[data-form-validation-target='passwordConfirmationError']"
    assert_select "div.hidden[data-form-validation-target='passwordStrength']"
  end
end